booking/src/main/java/org/dcsa/conformance/standards/booking/checks/BookingChecks.java

```diff
@@ -0,0 +1,645 @@
+    BookingState.DECLINED,
+    BookingState.REJECTED,
+    BookingState.CANCELLED,
+    BookingState.AMENDMENT_CANCELLED,
+    BookingState.AMENDMENT_DECLINED
+  );
+
+  private static final Set<BookingState> BOOKING_STATES_WHERE_CBR_IS_OPTIONAL = Set.of(
+    BookingState.RECEIVED,
+    BookingState.REJECTED,
+    BookingState.PENDING_UPDATE,
+    BookingState.UPDATE_RECEIVED,
+    /* CANCELLED depends on whether cancel happens before CONFIRMED, but the logic does not track prior
+     * states. Therefore, we just assume it is optional in CANCELLED here.
+     */
+    BookingState.CANCELLED
+  );
+  private static final JsonPointer CARRIER_BOOKING_REQUEST_REFERENCE = JsonPointer.compile("/carrierBookingRequestReference");
+  private static final JsonPointer CARRIER_BOOKING_REFERENCE = JsonPointer.compile("/carrierBookingReference");
+  private static final JsonPointer BOOKING_STATUS = JsonPointer.compile("/bookingStatus");
+  private static final JsonPointer[] TD_UN_LOCATION_CODES = {
+    JsonPointer.compile("/invoicePayableAt/UNLocationCode"),
+    JsonPointer.compile("/shipmentLocations/location/UNLocationCode"),
+    JsonPointer.compile("/shipmentLocations/location/facilityLocation/UNLocationCode")
+  };
+
+  public static ActionCheck requestContentChecks(UUID matched, Supplier<CarrierScenarioParameters> cspSupplier, Supplier<DynamicScenarioParameters> dspSupplier) {
+    var checks = new ArrayList<>(STATIC_BOOKING_CHECKS);
+    generateScenarioRelatedChecks(checks, cspSupplier, dspSupplier);
+    return JsonAttribute.contentChecks(
+      BookingRole::isShipper,
+      matched,
+      HttpMessageType.REQUEST,
+      checks
+    );
+  }
+  private static final JsonContentCheck CHECK_EXPECTED_DEPARTURE_DATE = JsonAttribute.customValidator(
+    "Check expected departure date can not be past date",
+    (body) -> {
+      String expectedDepartureDate = body.path("expectedDepartureDate").asText("");
+      var invalidDates = new LinkedHashSet<String>();
+      if(!expectedDepartureDate.isEmpty()) {
+        LocalDate expectedDepartureLocalDate = LocalDate.parse(expectedDepartureDate);
+        if (expectedDepartureLocalDate.isBefore(LocalDate.now())) {
+          invalidDates.add(expectedDepartureLocalDate.toString());
+        }
+      }
+      return invalidDates.stream()
+        .map("The expected departure date '%s' can not be past date"::formatted)
+        .collect(Collectors.toSet());
+    }
+  );
+
+  private static final JsonContentCheck CHECK_EXPECTED_ARRIVAL_POD = JsonAttribute.customValidator(
+    "Check expected arrival dates are valid",
+    (body) -> {
+      String providedArrivalStartDate = body.path("expectedArrivalAtPlaceOfDeliveryStartDate").asText("");
+      String providedArrivalEndDate = body.path("expectedArrivalAtPlaceOfDeliveryEndDate").asText("");
+      var invalidDates = new LinkedHashSet<String>();
+      if (!providedArrivalStartDate.isEmpty() && !providedArrivalEndDate.isEmpty()) {
```

Comment:
Does this catch the conditional check that if one is provided - then the other `MUST` be provided

Reply:
This should probably be a stand-alone check. There is a `allOrNoneArePresent` that is specialized for this.